cunumeric/coverage.py

```diff
@@ -59,26 +61,51 @@ def filter_namespace(
     }
 
 
-# todo: (bev) use callback protocol type starting with 3.8
-def implemented(func: Any, prefix: str, name: str) -> Any:
+class AnyCallable(Protocol):
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        ...
+
+
+class CuWrapped(Protocol):
+    _cunumeric_implemented: bool
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        ...
+
+
+def implemented(
+    func: AnyCallable, prefix: str, name: str, *, reporting: bool = True
+) -> CuWrapped:
     name = f"{prefix}.{name}"
 
-    @wraps(func)
-    def wrapper(*args: Any, **kwargs: Any) -> Any:
-        location = find_last_user_frames(not runtime.report_dump_callstack)
-        runtime.record_api_call(name=name, location=location, implemented=True)
-        return func(*args, **kwargs)
+    wrapper: CuWrapped
+
+    if reporting:
+
+        @wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            location = find_last_user_frames(not runtime.report_dump_callstack)
+            runtime.record_api_call(
+                name=name, location=location, implemented=True
+            )
+            return func(*args, **kwargs)
+
+    else:
+
+        wrapper = cast(CuWrapped, func)
```

Comment:
I had tried to use a "no-op" wrapper here just for consistency, but at least one test relies on being able to access ufunc attrs:
```
egate37 ❯ LEGATE_TEST=1 legate tests/reduction.py --gpus 2 -cunumeric:test                              
[0 - 7f52d5fff000]    1.667816 {6}{python}: python exception occurred within task:
Traceback (most recent call last):
  File "/home/bryan/work/legate.core/install37/lib/python3.7/site-packages/legion_top.py", line 410, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/home/bryan/work/legate.core/install37/lib/python3.7/site-packages/legion_top.py", line 234, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "tests/reduction.py", line 145, in <module>
    test()
  File "tests/reduction.py", line 39, in test
    r2 = num.add.reduce(x)
AttributeError: 'function' object has no attribute 'reduce'
```

I think this means that test would not pass if reporting is turned on.

Reply:
Nice catch. in case of a callable object, we should also wrap the methods with the reporting decorator. Since this is specific to `ufunc`, I guess we should just handle those objects.

Reply:
I think there are some questions/requirements to sort out so I have made a follow-on issue: https://github.com/nv-legate/cunumeric/issues/270 The current PR at least gets us as far as one place of responsibility for annotating the wrapper flag, so I'd like to merge it as-is if that is OK

Reply:
sounds good to me. let's make a follow-up PR for this issue.